subsys/bluetooth/services/cts_c.c

```diff
@@ -0,0 +1,251 @@
+
+	/* Reason for updating the time. */
+	p_time->adjust_reason.manual_time_update = (p_data[index] >> 0) & 0x01;
+	p_time->adjust_reason.external_reference_time_update =
+		(p_data[index] >> 1) & 0x01;
+	p_time->adjust_reason.change_of_time_zone = (p_data[index] >> 2) & 0x01;
+	p_time->adjust_reason.change_of_daylight_savings_time =
+		(p_data[index] >> 3) & 0x01;
+
+	return 0;
+}
+
+/**@brief Function for sanity check of a Current Time structure.
+ *
+ * @param[in] p_time  Current Time structure.
+ *
+ * @retval 0        If the time struct is valid.
+ * @retval -EINVAL  If the time is out of bounds.
+ */
+static int current_time_validate(current_time_char_t *p_time)
+{
+	/* Year. */
+	if ((p_time->exact_time_256.year > CTS_YEAR_MAX) ||
+	    ((p_time->exact_time_256.year < CTS_YEAR_MIN) &&
+	     (p_time->exact_time_256.year != 0))) {
+		return -EINVAL;
+	}
+
+	/* Month. */
+	if (p_time->exact_time_256.month > 12) {
+		return -EINVAL;
+	}
+
+	/* Day. */
+	if (p_time->exact_time_256.day > 31) {
+		return -EINVAL;
+	}
+
+	/* Hours. */
+	if (p_time->exact_time_256.hours > 23) {
+		return -EINVAL;
+	}
+
+	/* Minutes. */
+	if (p_time->exact_time_256.minutes > 59) {
+		return -EINVAL;
+	}
+
+	/* Seconds. */
+	if (p_time->exact_time_256.seconds > 59) {
+		return -EINVAL;
+	}
+
+	/* Day of week. */
+	if (p_time->exact_time_256.day_of_week > 7) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**@brief Function for handling of Current Time read response.
+ *
+ * @param[in] conn    Connection object.
+ * @param[in] err     ATT error code.
+ * @param[in] params  Read parameters used.
+ * @param[in] data    Pointer to the data buffer.
+ * @param[in] length  The size of the received data.
+ *
+ * @retval BT_GATT_ITER_STOP  Stop reading
+ */
+static uint8_t bt_gatt_cts_c_read_callback(struct bt_conn *conn, uint8_t err,
+					   struct bt_gatt_read_params *params,
+					   const void *data, uint16_t length)
+{
+	struct bt_gatt_cts_c *cts_c;
+	bt_gatt_cts_c_evt_t evt;
+	int err_code;
+
+	/* Retrieve CTS client module context. */
+	cts_c = CONTAINER_OF(params, struct bt_gatt_cts_c, read_params);
+
+	if (!err) {
+		err_code = current_time_decode(&evt.current_time, data, length);
+
+		if (err_code) {
+			/* The data length was invalid. Decoding was not completed. */
+			evt.evt_type = BT_GATT_CTS_C_EVT_INVALID_TIME;
+		} else {
+			/* Verify that the time is valid. */
+			err_code = current_time_validate(&evt.current_time);
+
+			if (err_code) {
+				/* Invalid time received. */
+				evt.evt_type = BT_GATT_CTS_C_EVT_INVALID_TIME;
+			} else {
+				/* Valid time received. */
+				evt.evt_type = BT_GATT_CTS_C_EVT_CURRENT_TIME;
+			}
+		}
+		cts_c->evt_handler(cts_c, &evt);
+	}
+
+	k_sem_give(&cts_c->read_params_sem);
+
+	return BT_GATT_ITER_STOP;
+}
+
+int bt_gatt_cts_c_current_time_read(struct bt_gatt_cts_c *cts_c,
+				    k_timeout_t timeout)
+{
+	int err;
+	struct bt_gatt_read_params *p_read_params;
+
+	err = k_sem_take(&cts_c->read_params_sem, timeout);
```

Comment:
I wonder if this semaphore is realy needed here. Maybe you could use atomic flag instead it ?
Then you can also remove `bt_gatt_cts_c_on_disconnected`

Reply:
Both semaphore and atomic flag protect the GATT read parameters variable in a similar way. When taking a semaphore, you may specify a timeout which allows waiting a little time. This timing convenience is not provided by atomic flag.

It is found that BLE disconnection does not reset semaphore state. bt_gatt_cts_c_on_disconnected() is written to handle this condition. The use of atomic flag still needs this handling.